applications/StructuralMechanicsApplication/custom_response_functions/custom_utilities/finite_differences_utilities.h

```diff
@@ -0,0 +1,477 @@
+            int direction_1 = 0;
+            int direction_2 = 0;
+            std::vector<Matrix> stress_vector;
+
+            if(item_1 == 'M')
+                rTracedElement.GetValueOnIntegrationPoints(SHELL_MOMENT_GLOBAL, stress_vector, rCurrentProcessInfo);
+            else if(item_1 == 'F')
+                rTracedElement.GetValueOnIntegrationPoints(SHELL_FORCE_GLOBAL, stress_vector, rCurrentProcessInfo);
+            else
+                KRATOS_ERROR << "Invalid stress type! " << traced_stress_type << (" is not supported!")  << std::endl;
+
+            if(item_2 == 'X')
+                direction_1 = 0;
+            else if(item_2 == 'Y')
+                direction_1 = 1;
+            else if(item_2 == 'Z')
+                direction_1 = 2;
+            else
+                KRATOS_ERROR << "Invalid stress type! " << traced_stress_type << (" is not supported!")  << std::endl;
+
+            if(item_3 == 'X')
+                direction_2 = 0;
+            else if(item_3 == 'Y')
+                direction_2 = 1;
+            else if(item_3 == 'Z')
+                direction_2 = 2;
+            else
+                KRATOS_ERROR << "Invalid stress type! " << traced_stress_type << (" is not supported!")  << std::endl;
+
+            unsigned int num_GP = stress_vector.size();
+            OutputVector.resize(num_GP);
+            for(size_t i = 0; i < num_GP; i++)
+            {
+                OutputVector(i) = stress_vector[i](direction_1, direction_2);
+            }
+        }
+        else
+        {
+            OutputVector.resize(1);
+            OutputVector.clear();
+        }
+
+        return OutputVector;
+
+        KRATOS_CATCH("")
+    }
+
+    double GetNodalDisplacement(const int NodeId, std::string TracedDofLabel, ModelPart& rModelPart )
+	{
+		KRATOS_TRY;
+
+        typedef Node<3>::Pointer PointTypePointer;
+        PointTypePointer traced_pNode = rModelPart.pGetNode(NodeId);
+
+        typedef VariableComponent<VectorComponentAdaptor<array_1d<double, 3>>> VariableComponentType;
+		const VariableComponentType& rTRACED_DOF =
+            KratosComponents<VariableComponentType>::Get(TracedDofLabel);
+
+		double displacement_value = traced_pNode->FastGetSolutionStepValue(rTRACED_DOF);
+
+		return displacement_value;
+
+		KRATOS_CATCH("");
+	}
+
+    double GetStrainEnergy( ModelPart& rModelPart )
+    {
+        KRATOS_TRY;
+
+        ProcessInfo &CurrentProcessInfo = rModelPart.GetProcessInfo();
+		double strain_energy = 0.0;
+
+		// Sum all elemental strain energy values calculated as: W_e = u_e^T K_e u_e
+		for (ModelPart::ElementIterator elem_i = rModelPart.ElementsBegin(); elem_i != rModelPart.ElementsEnd(); ++elem_i)
+		{
+			Matrix LHS;
+			Vector RHS;
+			Vector u;
+
+			// Get state solution relevant for energy calculation
+			elem_i->GetValuesVector(u,0);
+
+			elem_i->CalculateLocalSystem(LHS,RHS,CurrentProcessInfo);
+
+			// Compute strain energy
+			strain_energy += 0.5 * inner_prod(u,prod(LHS,u));
```

Comment:
mmm
this is only true for a linear element. 

You should call the Calculate function of the element for it being done properly